[HTTP] Increase timeout on send

clumsy-stefan opened the pull request:
Increase timeout for sending HTTP chunks.

clumsy-stefan committed:
[HTTP] Increase timeout on send

Increase timeout for sending HTTP chunks.

clumsy-stefan commented:
Not sure if this is a real issue, but I found especially on nodes with low RSSI that the connection dropped during the transfer of data. Same happened when getting /json information.

After increasing the timeout on chunked transfers the issue disappeared (or at least happened much less).

I think especially the default timeout of 1 on line 326 could cause issues. If I understand this correctly, the timeout is set to this value if there is enough memory (which probably happens a lot), but 1ms timeout is probably not enough.

I'l keep it running on my nodes now and see if it fixes my issues on the long run.

TD-er commented:
The default of 1 is indeed a bit low, but on the other hand, it should be an open connection...
I guess it doesn't hurt to make it longer.

clumsy-stefan commented:
I'm not sure what reasonable values are and if it could also be too high?

Any idea what else could cause these partial transfers I observe on loaded WiFi networks and/or low RSSI?

tonhuisman commented:
> and if it could also be too high?

The essence of a time-out value is that it's a maximum, so if it completes before it expires, it will continue as intended 😉 
If you make it too long and it happens often, that could be annoying for the user as it might slow things down, but as said, 1 msec is rather short. Failing connections are way more annoying though.

clumsy-stefan commented:
Why is that value made dependent on the amount of free memory and not on the quality of the connection?

TD-er commented:
> Why is that value made dependent on the amount of free memory and not on the quality of the connection?

Clearing up memory takes time. e.g. flushing to some controller, or pending buffers which are being processed in the background.
Also memory allocation takes more time when the memory gets more fragmented.

clumsy-stefan commented:
Changing the timeout(s) seems to improve the reliability of the connections of my nodes, but does not completely eliminate the partial transfers. But it's way better than with the smaller timeouts. So I guess it would not do any harm to include these changes IMHO.

tonhuisman commented:
Does the reliability improve if you increase these time-outs further?

clumsy-stefan commented:
I'll try that. Just did a new version with doubled timeouts, will  test that today.

TD-er commented:
What kind of connection do you use to access the ESPs?
Is it maybe over some 4G/VPN or perhaps a bit crowded ether due to lots of access points?

clumsy-stefan commented:
~40 Nodes distributed over 3 floors with 3 Mikrotik AP's. But lots of traffic also due to other Equipment. So for sure it's not a perfect environment, but like this at least you can fine tune such issues ;) In an ideal world they would never occur... Probably we can make it a bit more robust like this..

TD-er commented:
> So for sure it's not a perfect environment

More like a realistic environment :)

clumsy-stefan commented:
> Does the reliability improve if you increase these time-outs further?

First observations suggest that it get's worse again...

I did a number of different versions. Currently I would say that setting the default timeout to 200-300ms and remove the memory tests therefore not changing the timeout any more has the best impact for me... however these are very subjective and heuristic findings, also quite short-timed view... 

The inly thing I'm quite sure is, that the default timeout should not be only 1ms... the rest seems very dependent on the current traffic on the wifi network... 

I'll keep the version with a default of 300ms and no memory dependent change running for now and see how it runs over night...

> More like a realistic environment :)

I would also say so, and together with the neighbours a real mess in the 2G frequency band...

TD-er commented:
> I would also say so, and together with the neighbours a real mess in the 2G frequency band...

A friend of mine lives in a 12-story appartment building, so he has at least all APs of his neighbours in view withing 3 appartments left/right and 3 floors up and down.
There are several of these standing in a row, facing the balcony sides, so he also has all APs of both buildings on each side in view.
That's about 150+ APs.
He can't make a Skype/meet call using 2.4 GHz WiFi, without disconnects.
So I know how bad things can get on the 2.4 GHz band :)

clumsy-stefan commented:
I'm wondering if playing with the TCP_MSS and CHUNKED_BUFFER_SIZE would also have an impact...

TD-er commented:
> I'm wondering if playing with the TCP_MSS and CHUNKED_BUFFER_SIZE would also have an impact...

I did play with the buffer size only a few weeks ago and the chunked buffer size is fine for ESP8266.
The TCP_MSS I didn't touch, but I got the impression this was way lower than 1400-ish on ESP8266.
More like 500-ish bytes. Therefore it is best to have the buffer size at 400 as it is now, otherwise you may end up with almost double the amount of packets for the same data when you just approach the MSS value with the buffer size.

clumsy-stefan commented:
I normally compile with an MSS of 536 optionally I could select 1460, but then it obviously uses a bit more memory...

CHUNKED_BUFFER_SIZE has a default of 512 in `Web_StreamingBuffer.cpp`. Probably these two are too close together?

TD-er commented:
> CHUNKED_BUFFER_SIZE has a default of 512 in `Web_StreamingBuffer.cpp`. Probably these two are too close together?

Yep, could be.
Have to check the size of the TCP header

TD-er commented:
Looks like this header may be 24 bytes, which makes a total of exactly 536.
But if there are some other headers being wrapped, you may end up with funky issues indeed.
Better set it to 480-ish (or 400 as it was)

clumsy-stefan commented:
hmm.. but why make it that small and not leave the fragmenting to the tcp-stack?

You can anyway never be sure what the MSS size is or even what the packet size (MTU from the AP) really is... 

I'm just trying the opposite, use a big chunk buffer and leave the rest to the network stack.

TD-er commented:
The problem with leaving it to the network stack is that whenever a package doesn't get acknowledged, your 'penalty' is worse as those packets may be kept in memory for quite some time.
For example, if your ESP reconnects, then those packets will never get acknowledged and it may take 5 - 10 minutes before LWIP frees this memory.

The MTU of the rest of the network is probably not an issue as I hardly ever see such small MTU/MSS values.
Usually the MTU is something like 1492 or 1500 bytes and the MSS is usually 1300 - 1400-ish 

You can test what the max. MTU is by sending ping packets to the ESP node with the 'do not fragment' flag set and a size.
This way you can see if the actual size is less than the expected size, which is then often caused by some wrapping of packets. For example VPN routing.

N.B. this 536 value as build flag, is this the MSS or the MTU value? I hope it is just the MSS.

clumsy-stefan commented:
I also use an MTU of 1500 on my network (generally) which losses a few bytes due to the WiFi overhead, so it's 1496 on the physical network.

It's only MSS as a build flag, I don't touch the MTU of the ESP core.

I'll need to test with an MSS of 1460 and set the buffer to 1024 or such, probably that helps as it would need to send less individual wifi packets and thus use less wifi bw... (I think)...

clumsy-stefan commented:
Somehow I get the feeling, that increasing the timeout further makes the node even more unresponsive.

Actually I'm not even sure if I do understand the meaning of this timeout value correctly, when reading the code. What actually happens if this timeout expires? Does it stop sending? If I understand that part coreectly it just loops and waits until either the timeout expires or enough RAM is there to allocate a new buffer. So theoretically it could be, that a higher timeout stalls the node (and all other operations), also as it seems to use blocking send?

TD-er commented:
https://github.com/letscontrolit/ESPEasy/blob/c5a87a35adc3eb3851c9987e1090cc1bf7008a03/src/src/DataStructs/Web_StreamingBuffer.cpp#L326-L346

The data has already been sent.
The timeout is the amount of time spent on trying to allocate memory for the send buffer, or trying to free RAM.
During this timeout period, `delay()` is called, which does also perform some background tasks, like flushing buffers.
But it also allows to receive data being sent to the ESP.
So in the end a longer timeout may result in even more memory being used.

Thus increasing the max timeout is probably not going to help and in some cases making it even worse as the incoming packets are not being served.

I do agree that the minimal timeout of 1 is a bit low as it only allows for 1 attempt. But increasing the other 2 values is probably useless.
I think the minimal timeout of 100 is probaby just fine, as that's the WiFi beacon interval time, which may be just what's needed to flush some data.

clumsy-stefan commented:
> So in the end a longer timeout may result in even more memory being used.

I'm not sure if memory is that critical anymore. After all the improvements you have done in the past I rarely observe critical memory values

> Thus increasing the max timeout is probably not going to help and in some cases making it even worse as the incoming packets are not being served.

Yep, I guess that's what I observed

> I think the minimal timeout of 100 is probaby just fine, as that's the WiFi beacon interval time, which may be just what's needed to flush some data.

I currently run it with 300ms, and expierence interrupts and missing ICMP responses on nodes with low RSSI and/or heavy load. So I'll do a new version with 100ms to see what's happening...

TD-er commented:
ICMP response is not really related to this, as pings are being dealt with during `delay` calls.
For those, you should have to look into the timing stats, to see if there is something blocking for quite a while.

clumsy-stefan commented:
I "feel" that I get best results with a fixed timeout of 200ms...

The other problem seems to be the interaction between my AP's and the nodes... especially after restarting the nodes it takes some time until the connection becomes stable. I still have get packet loss from time to time, I think particularly when the wifi gets crowded... need to track that down further...

TD-er commented:
> I still have get packet loss from time to time, I think particularly when the wifi gets crowded... need to track that down further...

Timingstats :)

clumsy-stefan commented:
> Timingstats :)

Yep, but that's not that easy... Especially on the nodes where (and when) the issue raises I only get partial pages, mostly only the header lines and no usable contents... Intrestingly it seems to get better the longer the nodes run...

However I also get it on nodes where I have only one task (sysinfo) and nothing else configured... It really seems dependent on the Network. There I tried a dozen of different low level settings on the AP (like frame-timeouts, noise adaptivenes, etc.).

Currently I can't see any pattern though...

TD-er commented:
Can you see on your access point at what RSSI the ESP is received by the AP?
If you compare this with the reported RSSI on ESPEasy, does it differ a lot?
Is there enough room left to lower the max. TX power in ESPEasy? (perhaps also enable to always send with max. TX power, but just limit it)
See here for more hints and tips on TX power and sensitivity margin: https://espeasy.readthedocs.io/en/latest/Tools/Tools.html?highlight=TX%20power#wifi-tx-power

Also do not enable "ECO" mode if you see packet loss and do enable sending Gratuituous ARP packets.
Is your unit connecting using 'g' mode or 'n' mode?
If it is 'n' mode, which SDK flags are you using in your build?
Can you also try building using the latest "beta" code, as 3.1.1 is now released and I got the impression they are now using a different default "binary blob" for the WiFi.
Ran it here on 1 node and it was working just fine.
However those builds using the latest SDK/Arduino version tend to be 5 - 10k larger.
Change in memory usage can be both positive and negative, highly depending on your use case.

TD-er commented:
Oh and if you have some idea which plugin may cause this heavy memory use, please let me know. Maybe some extra TLC on that code can be useful too :)

clumsy-stefan commented:
First of all, thank you very much for taking the time and helping me and suggesting ways forward, I relly appreciate and don't see this as "normal"!!

> Can you see on your access point at what RSSI the ESP is received by the AP?

yes

> If you compare this with the reported RSSI on ESPEasy, does it differ a lot?

usually the node tells me around 2-3dB less

> Is there enough room left to lower the max. TX power in ESPEasy? (perhaps also enable to always send with max. TX power, but just limit it)

All my nodes have TX enabled and Max TX Power 17.5dB.

> Also do not enable "ECO" mode if you see packet loss and do enable sending Gratuituous ARP packets.

ECO is disabled and GARP on all enabled (however the ECO mode didn't seem to make any difference, I had the same issue before, but just to make sure I disabled it a month ago or so).

> Is your unit connecting using 'g' mode or 'n' mode?

all in N-Mode

> Can you also try building using the latest "beta" code, as 3.1.1 is now released and I got the impression they are now using a different default "binary blob" for the WiFi.

I swithed back to NONOS 2.2.1 as it seemed more stable to me, but I can try with a version 3.

> Ran it here on 1 node and it was working just fine.

Yes running (only) one node for a few minutes/hours is mostly not an issue. I really think it's the sheer mass of wifi clients that starts to interfere and make a mess in the communication.

> However those builds using the latest SDK/Arduino version tend to be 5 - 10k larger.

thats one of the reasons I switched back.

> Oh and if you have some idea which plugin may cause this heavy memory use, please let me know. Maybe some extra TLC on that code can be useful too :)

sure, whatever I find I'll report... unfortunately it also happens on nodes with just one task (sysinfo) enabled...

After updating all my nodes with a new version, it usually takes between 15min. and some hours until everything has settled again. I monitor the nodes via LibreNMS (just ping), Zabbix (getting the /json page and ICMP) and the values are sent with the FHEM controller to my server. Especially on Zabbix where I grab the json every 5min. I can see that the first 15-60min. It rarely can fetch the whole page, after some hours it works more or less stable with a failed page every now and then. Both systems are happy with ping, as they this is less critical (just 1 -3 packets sent and received). That's why I think the interaction with the AP's and the rest of the network also is an issue....

TD-er commented:
> All my nodes have TX enabled and Max TX Power 17.5dB.
> 

That's not what I meant.
I mean, is the RSSI such that it has some room to lower it like 10 dBm?
Just try to keep > -70 dBm, so if you now are connecting with an RSSI of -50dBm, there is roughly 20 dBm margin to play with.

When connecting using 802.11n, the ESP can't send with more than 14 dBm, so you could set the max. TX power to 4 dBm and expect an RSSI of 10 dBm less.

When using 802.11g, you have a bit more room to play with regarding the range of acceptable RSSI values.
Also there seems to be some 'issue' with 802.11n and ESP8266 as most "binary blobs" later than March 2019 seem to be not 100% compatible with the 802.11n standard. That's why I asked about what SDK build you're using in your builds.
Some APs seem to purposely disconnect nodes connecting via 'n' every now and then and it seems as if the ESPs don't handle this very well. When using 'g' mode, this does not happen like for 'n' connected devices.

When fetching the JSON (I assume) via Zabbix, do you just have some URL entered in Zabbix, or are you calling a script?
If calling a script, please try to send out 1 ping right before and see if this makes a difference.

clumsy-stefan commented:
> I mean, is the RSSI such that it has some room to lower it like 10 dBm?

Obviously with 40 nodes I have all different kind of RSSI's usually between -50 and -80. But the same thing happens on nodes with -50 and on nodes with -80. I have nodes that are rock-stable at -78...

> When using 802.11g, you have a bit more room to play with regarding the range of acceptable RSSI values.
Also there seems to be some 'issue' with 802.11n and ESP8266 as most "binary blobs" later than March 2019 seem to be not 100% compatible with the 802.11n standard. That's why I asked about what SDK build you're using in your builds.
Some APs seem to purposely disconnect nodes connecting via 'n' every now and then and it seems as if the ESPs don't handle this very well. When using 'g' mode, this does not happen like for 'n' connected devices.

Now that's an interesting one. Some time ago (1-2 years) I used to run all nodes in fixed G mode as it seemed more stable to me. After you did some changes in the code (probably you remember the intermittent disconnects and not beeing able to reconnect issue) I switched back to N as it seemed completely stable again. But since then Core versions and also AP-Firmwares changed a number of times obviously.... Probably I will change back all nodes again to G Mode and see if I can spot a difference. Changing only one node doesn't make sense, as the issue occurs randomly over all nodes.

> When fetching the JSON (I assume) via Zabbix, do you just have some URL entered in Zabbix, or are you calling a script?
If calling a script, please try to send out 1 ping right before and see if this makes a difference.

Directly the URL from Zabbix. But as I also have the SNMP profile assigned I guess it anyway does an ICMP Ping before fetching the json.

Running a constant ping to the more critical nodes also shows every 10th or so packet is lost....

TD-er commented:
> Obviously with 40 nodes I have all different kind of RSSI's usually between -50 and -80. But the same thing happens on nodes with -50 and on nodes with -80. I have nodes that are rock-stable at -78...

The reason I suggest to lower the TX power is that on some nodes it seems as if the reflected RF power from a badly tuned antenna is interfering with the WiFi radio itself.
With less TX power, obviously less power is reflected.

Also lower TX power is a bit easier on the power supply components.

clumsy-stefan commented:
Ok, but setting it to 4dBm seems very low to me?

The problem is if I can't reach the node at all anymore, it gets quite annoying accessing the nodes again (as some of them are not that easy accessible)...

What's the advantage of using N mode instead of G? If G is more stable, do I loose anything if I force it to G again?

TD-er commented:
> Ok, but setting it to 4dBm seems very low to me?
> 

That's why you should check the 'margin' you have.
If you uncheck the "send with max TX power", ESPEasy does alter the TX power automatically as I described.
But then you must also set the sensitivity margin as the difference between RSSI on the AP and on the ESPEasy node. (typically 3 dB)

> The problem is if I can't reach the node at all anymore, it gets quite annoying accessing the nodes again (as some of them are not that easy accessible)...
> 

Yep, that's indeed correct.
I don't assume you immediately test these scenarios on 100s of nodes, starting with those which need breaking away wall panels to access  if you can't reach them ;)

> What's the advantage of using N mode instead of G? If G is more stable, do I loose anything if I force it to G again?

The advantage of 'n' mode on an ESP node isn't really obvious as you simply never going to need the extra bandwidth.
However quite a lot of access points seem to have difficulties when the need to switch between modes (b/g/n)
And for other devices like laptops etc. you can benefit from the 'n' mode over 'g' as it allows for more bandwidth and also better manages areas which have lots of devices.
But if you only have slow devices connected to an AP, I would set it to 'g' mode for sure.

clumsy-stefan commented:
> If you uncheck the "send with max TX power", ESPEasy does alter the TX power automatically as I described.
But then you must also set the sensitivity margin as the difference between RSSI on the AP and on the ESPEasy node. (typically 3 dB)

I would need to do this for each node individually? And depending on which AP the node actually connects to (some nodes see more than one AP) the RSSI (and threfore margin) will be different....

> I don't assume you immediately test these scenarios on 100s of nodes, starting with those which need breaking away wall panels to access if you can't reach them ;)

Yes and no... If I only change one or two nodes it's not really that valuable, as single nodes always run fine, especially the test nodes wich usually are easily accessible. and because they are easy accessible they mostly also have a good RSSI ;)
So I'll mostly take the risk and hope I don't need to dig out the screwdriver.. also because I'm currently doing the changes remotely 😄 

> But if you only have slow devices connected to an AP, I would set it to 'g' mode for sure.

I think I'm going to try this next..

TD-er commented:
> I would need to do this for each node individually? And depending on which AP the node actually connects to (some nodes see more than one AP) the RSSI (and threfore margin) will be different....

Nope, you might want to test a bit to see if there are significant differences between RSSI on both ends.
Of course, only compare the ones which actually are connected to eachother :)
Typically this difference should be about 3dB.
If this difference is (much) more, I would consider replacing the ESP node as the antenna is probably badly tuned, or something is obstructing the antenna.

clumsy-stefan commented:
I begin to think that it's not a problem of RSSI but latency and/or other timeouts (probably layer 2)... probably the wifi is just too crowded that to may hw-frames get lost and the HW retries (currently 7) are missed at some point in time. Could also be some incompatibility between my AP's (Mikrotik) and the esp, as I don't see this behaviour on any other (wifi) HW in my net (and I have ~100 devices talking wireless)... even the cameras that stream via wifi run smoothly without interrupts...

While I really appreciate your help and advise I think the issue is more fundamental and probably not even related to ESPEasy. I'll continue to try different setups and configs, but somehow I need to find a different approach as tweaking around wifi settings on the node or on the AP's  just doesn't really change the behaviour. Still, if you have any more pointers where to look for I appreciate though, but doesn't want to overuse your time ;)

TD-er commented:
Mikrotik allows to set some profile, like "indoor" and "outdoor".
Please set it to the latter, as this allows for some longer timeouts.

clumsy-stefan commented:
thanks, sure, one of the first things I tried (indoor, dynamic, 1, 10, 100km...) also tried to tweak all the other HW settings (as I'm a network engineer I'm better at network settings than programming, that's why I use MT's... we use them widely for our network in our company as we're running a broadband network and datacenter ;) That's also why I think it's the WiFi part...

clumsy-stefan commented:
PS: also interesting, sending via controller 9 / FHEM I have absolutely no issues and data is transmitted correctly, so only when needing to do chunked transfers (more than 512B of data) the issues occur....

TD-er commented:
Ah, I also am not a stranger in datacenters and with rather non-consumer networking :)
I also like the Mikrotik stuff, as it is just the perfect price/performance ratio. 
But it is a rather steep learning curve to get along with those. The first time I used one 20-euro MikroTik unit, I kinda felt lost as I really had to search for quite a lot of terms.
You just don't expect BGP config in a 20-euro device, so I had to make sure it wasn't something else which I didn't know :)

One other thing you could check is to see whether you have a non-standard DTIM or beacon interval.
The default beacon interval is 100 units (102.4 msec).
I don't know whether Mikrotik may have some parameter for how many beacon intervals it may keep the packets before considering a timeout?
Is there some device in your network set to act as an ARP proxy server?

Do you have multiple SSIDs set to the same MikroTik AP? (on the same band that is)

TD-er commented:
![image](https://user-images.githubusercontent.com/3751318/220434956-0441de61-394e-458e-9fc4-bc2082a8fc34.png)
This is how I have one of my MikroTik units configured

Some of the "Advanced settings"
![image](https://user-images.githubusercontent.com/3751318/220435346-9755eee7-37e7-4d3b-a694-f7249463fefc.png)


Maybe you should deselect the "Supported Rates B"

clumsy-stefan commented:
absolutely agree... but after tenish or so years you get used to it... 

> I don't know whether Mikrotik may have some parameter for how many beacon intervals it may keep the packets before considering a timeout?

Yep tried that, adjusting timings and also number of failed frames etc. the best thing I managed to do though is that most of the other devices were not able to connect to the AP's anymore (as well as the ESP's ;)

> Is there some device in your network set to act as an ARP proxy server?

no. Proxy-Arp is kind of a bad thing to do and you really need to make sure you have it on the right interfaces. Besides this would also impact all the other devices.

> Do you have multiple SSIDs set to the same MikroTik AP? (on the same band that is)

Yes, normally 2 SSID's on all AP's and always on both bands (2G & 5G) sepparated by VLAN's, all MT's via tagged VLAN bridges connected (3 AP's 2small and one CCS switch and a CCR as firewall). However the main SSID where also the ESP's are are directly on the HW card (not a virtual AP) and without VLAN tag....

TD-er commented:
> Yes, normally 2 SSID's on all AP's and always on both bands (2G & 5G) sepparated by VLAN's, all MT's via tagged VLAN bridges connected (3 AP's 2small and one CCS switch and a CCR as firewall). However the main SSID where also the ESP's are are directly on the HW card (not a virtual AP) and without VLAN tag....

I have done this too a while ago and stopped having >1 SSID per Mikrotik as it did cause strange issues.
So can you test with only 1 SSID per Mikrotik?

VLAN tagging shouldn't really matter as that's done in the switch port, not in software.

clumsy-stefan commented:
![image](https://user-images.githubusercontent.com/32835100/220436361-09517ac7-7e60-4ef7-a9ea-e1814cb8132e.png)
Looks similar. The main difference is probably the Bands and Channel widths..

clumsy-stefan commented:
> I have done this too a while ago and stopped having >1 SSID per Mikrotik as it did cause strange issues.
So can you test with only 1 SSID per Mikrotik?

We use this in our company with a multitude of SSID's and have no issues, also all other devices don't have issues...

> VLAN tagging shouldn't really matter as that's done in the switch port, not in software.

Depends if you use the "old" VLAN tagging as it was in version up to 6.something I guess where it's done on the physical HW switch except for WiFi and other virtual interfaces, where its done by the CPU  or if you use the "new" way of tagging where everything is configured in the bridge (with vlan tagging on). Newer MT's (as the CCS) don't even have the switch part anymore, everything down to HW-Offloading is done on the bridge and port configs.

clumsy-stefan commented:
can a node handle multiple simultaneous  http requests? Could it be an issue when 2 different clients access the node more or less at the same time? Could some memory mix up happen?

it's really the chunked transfers which cause issues... when it only needs to send 1 packet I see no problems... (just FYI..)

TD-er commented:
> can a node handle multiple simultaneous http requests

Nope, not really.

clumsy-stefan commented:
> https://github.com/letscontrolit/ESPEasy/blob/c5a87a35adc3eb3851c9987e1090cc1bf7008a03/src/src/DataStructs/Web_StreamingBuffer.cpp#L326-L346

I'm wondering if 1) that part is really usefull at all (I have a running version without) and 2) if the `data.reserve()` in the check could cause a problem. To me it seems that this is looping and reserving more and more memory until mem is <4000 and/or timeout is reached. Probably as the original timeout was 1 this never happened as it run only once through that loop.
EDIT: this obviously only happens once mem is below 4000 but is still able to allocate the buffer...

However I'm not sure what exactly `data.reserve()` does, but couldn't that produce a memory leak, as it's never freed again?

TD-er commented:
Nope, it only tries to reserve the given amount of memory, no more, no less.
The call to `data.clear()` at least deletes the content of the string. It currently does not free the amount of memory, so a call to `data.reserve` essentially doesn't do anything right now.
But I will leave it there, as there are plans to change the functionality of `clear()` in that class to also free the allocated memory.

clumsy-stefan commented:
Ok, and having it reserved doesn't mean it's reduced from the free mem? So basically you can call `data.reserve()` in a loop an never run out of memory?

TD-er commented:
> Ok, and having it reserved doesn't mean it's reduced from the free mem? So basically you can call `data.reserve()` in a loop an never run out of memory?

When calling it with the same integer value and given it already has this amount allocated, then yes, you can call it forever.

```c++
bool String::reserve(unsigned int size) {
    if(buffer() && capacity() >= size)
        return true;
    if(changeBuffer(size)) {
        if(len() == 0)
            wbuffer()[0] = 0;
        return true;
    }
    return false;
}
```

TD-er commented:
There is only 1 catch... What if for whatever reason, this buffer outgrew its max. allocated size.
Then it would not be reduced.
However I don't think anywhere in the code it can get larger than the intended size.

You can add a check for this at the line with `data.clear(); `

```c++
if (data.length() > CHUNKED_BUFFER_SIZE) {
  data = String(); // Clear also allocated memory
} else {
  data.clear();
}
```

clumsy-stefan commented:
Ok, thanks, I'll try this...

It just seems, that when the network is slow or has big latency (eg. due to crowded air) the node can't get rid of the packets and looses them. Could also be in the core though...

TD-er commented:
Could also be in the opposite direction, where the acknowledgement isn't received by the ESP.
Since it is TCP and with chunked transfer you do have quite a lot of packets, it may get stalled.

LWIP (the IP stack) also needs to allocate memory to store the received packets.
Whenever this fails, it may get stuck as it will not continue handling the ongoing transfer which may free up memory when done.

But maybe we should also look into why your nodes are running so incredibly low on memory in the first place.

clumsy-stefan commented:
My nodes are normally over 10k free mem (way over the 4k for sure), so I see the problem more in something as you describe, like missing an ack or similar than in low mem...

<img width="823" alt="image" src="https://user-images.githubusercontent.com/32835100/220560847-673970cc-c471-49b1-976a-29c3a6afe010.png">

TD-er commented:
Maybe we can tweak some timeout settings in the webserver component, but I'm not sure if that will make things better or worse.
It seems like the timeout to wait for an ack. might be too long.
But perhaps it also depends on the used AP, as not all APs keep retrying packets for the same amount of time.
That's why I mentioned DTIM settings, as those often also affect any such retry timeout settings in the AP.

clumsy-stefan commented:
I agree. On the other hand having ~100 WiFi devices running on these AP's and only the ESP8266 having problems, I'm a bit cautious on changing AP settings... Also my ESP32's (~5) never have/had these issues.... Even if they have bad RSSI. That's why I'm more thinking about the esp core (or lwip) part....

TD-er commented:
Sure both are quite different platforms.
But still there is one obvious difference.
The ESP8266 does have a lot less memory and if this gets fragmented, memory allocation needs more time to find a block large enough to hold the needed amount.
On the ESP32, memory fragmentation does affect performance a lot less as the size of the needed blocks doesn't really change, so it is more likely memory allocation can be much faster.

Also the WiFi code is running in a separate RTOS task on the ESP32 familiy of nodes.
On the dual core nodes, I think it even runs on a different core.
But since it is running RTOS, it has a much more guaranteed interval where it is being served.
On the ESP8266 any so called "blocking code" may interfere with the WiFi operations.
On the timingstats page I mark anything > 100 msec in bold, but essentially everything running > 10 msec may already affect WiFi performance.
If a unit is also running some plugins which need to disable interrupts for some time (e.g. Dallas 1Wire, some DHT sensors, IR, SW serial) then all bets are off.

clumsy-stefan commented:
probably not related, but one of the test nodes where I use for above changes just rebooted (after not beeing really responsive) and now tells me:
<img width="566" alt="image" src="https://user-images.githubusercontent.com/32835100/220566508-f6324fe8-7581-4e79-b5eb-d7ae3d36b7c5.png">

But I only have C009 (FHEM) enabled and no MQTT of any sort?! Why does it go through `TIMER_C006_DELAY_QUEUE`?

TD-er commented:
I have no clue, have to check it later...

TD-er commented:
Ah, found it :)
The string is incorrect.
It takes the enum value and converts it to an int.
But the enum doesn't have all C001....C025 defined.

```c++
 TIMER_MQTT_DELAY_QUEUE,
    TIMER_C001_DELAY_QUEUE,
    TIMER_C003_DELAY_QUEUE,
    TIMER_C004_DELAY_QUEUE,
    TIMER_C007_DELAY_QUEUE,
    TIMER_C008_DELAY_QUEUE,
    TIMER_C009_DELAY_QUEUE,
    TIMER_C010_DELAY_QUEUE,
```
So `enumValue - TIMER_C001_DELAY_QUEUE` isn't equal to the number in this enum string.

clumsy-stefan commented:
do you know what LWIP options you use when you compile with PIO?
Normally I use:
`-DLWIP_OPEN_SRC -DTCP_MSS=536 -DLWIP_FEATURES=1 -DLWIP_IPV6=0`

EDIT:
The reason behind the question is, that I see for problematic nodes all kind of strange things happening when sniffing the packets:
<img width="1556" alt="image" src="https://user-images.githubusercontent.com/32835100/220744322-222a6e5d-ffdc-48ec-98be-26482df61051.png">
Googling around a bit, I found, that these retransmission and RST ACK's seemed to be a known problem of LWIP stack. Now I'm wondering if it could be fixed with the right LWIP options...

TD-er commented:
> do you know what LWIP options you use when you compile with PIO?

See:
https://github.com/letscontrolit/ESPEasy/blob/mega/platformio_core_defs.ini#L59

clumsy-stefan commented:
AS I do not know PIO where is `PIO_FRAMEWORK_ARDUINO_LWIP2_HIGHER_BANDWIDTH_LOW_FLASH` translated to the LWIP Options as defined in [lwipopts](https://github.com/esp8266/Arduino/blob/master/tools/sdk/lwip2/include/lwipopts.h), can you point me?

TD-er commented:
```python
#
# lwIP
#
lwip_lib = None
if "PIO_FRAMEWORK_ARDUINO_LWIP2_IPV6_LOW_MEMORY" in flatten_cppdefines:
    env.Append(
        CPPDEFINES=[("TCP_MSS", 536), ("LWIP_FEATURES", 1), ("LWIP_IPV6", 1)],
        CPPPATH=[join(FRAMEWORK_DIR, "tools", "sdk", "lwip2", "include")],
    )
    lwip_lib = "lwip6-536-feat"
elif "PIO_FRAMEWORK_ARDUINO_LWIP2_IPV6_HIGHER_BANDWIDTH" in flatten_cppdefines:
    env.Append(
        CPPDEFINES=[("TCP_MSS", 1460), ("LWIP_FEATURES", 1), ("LWIP_IPV6", 1)],
        CPPPATH=[join(FRAMEWORK_DIR, "tools", "sdk", "lwip2", "include")],
    )
    lwip_lib = "lwip6-1460-feat"
elif "PIO_FRAMEWORK_ARDUINO_LWIP2_HIGHER_BANDWIDTH" in flatten_cppdefines:
    env.Append(
        CPPDEFINES=[("TCP_MSS", 1460), ("LWIP_FEATURES", 1), ("LWIP_IPV6", 0)],
        CPPPATH=[join(FRAMEWORK_DIR, "tools", "sdk", "lwip2", "include")],
    )
    lwip_lib = "lwip2-1460-feat"
elif "PIO_FRAMEWORK_ARDUINO_LWIP2_LOW_MEMORY_LOW_FLASH" in flatten_cppdefines:
    env.Append(
        CPPDEFINES=[("TCP_MSS", 536), ("LWIP_FEATURES", 0), ("LWIP_IPV6", 0)],
        CPPPATH=[join(FRAMEWORK_DIR, "tools", "sdk", "lwip2", "include")],
    )
    lwip_lib = "lwip2-536"
elif "PIO_FRAMEWORK_ARDUINO_LWIP2_HIGHER_BANDWIDTH_LOW_FLASH" in flatten_cppdefines:
    env.Append(
        CPPDEFINES=[("TCP_MSS", 1460), ("LWIP_FEATURES", 0), ("LWIP_IPV6", 0)],
        CPPPATH=[join(FRAMEWORK_DIR, "tools", "sdk", "lwip2", "include")],
    )
    lwip_lib = "lwip2-1460"
# PIO_FRAMEWORK_ARDUINO_LWIP2_LOW_MEMORY (default)
else:
    env.Append(
        CPPDEFINES=[("TCP_MSS", 536), ("LWIP_FEATURES", 1), ("LWIP_IPV6", 0)],
        CPPPATH=[join(FRAMEWORK_DIR, "tools", "sdk", "lwip2", "include")],
    )
    lwip_lib = "lwip2-536-feat"

```

clumsy-stefan commented:
Ok, interesting, so you use MSS of 1460 and no features (meaning no fragmentation and no OOSQ handling...). Need to try that.... Thanks!!!

clumsy-stefan commented:
Just another observation:

After another 10 or so versions with different parameters I found one other "subjective" behaviour. Depending on the settings I use in the FHEM-Controller (C009) for timeouts, retries, etc. I get the interrupted transfer more or less often. It seems to me, that these transfers interfere with each other. Probably a background timing issue, if the controller sending blocks for too long, the web page can't be served, or the other way round. Especially on a loaded network or if the clien/fhem/browser takes a bit longer to respond or BW is low...

TD-er commented:
That's rather strange and suggests that buffer allocation fails, but not being noticed as the pointer still points to some data previously being processed.
Or... a heap allocated object is being processed as a char pointer and not the data, but rather the pointer is being added to some processing queue.
Both are not OK, but I'm not sure these can be fixed or properly detected in ESPEasy itself as that's an internal problem in either LWIP or the WebServer class.

Have to think about this, as this might be explaining all those issues we're seeing, but dealing with them will be quite hard.

clumsy-stefan commented:
Ok. I'm not sure how esp handles background operation, but couldn't it also be, that the lower lwip stack just doesn't get enough interrupts/calls/whatever to actually handle network traffic? As I also see missing ICMP responses even though it seems the node is not loaded but has some latency in network communications..

TD-er commented:
That's why I started collecting timing stats, to pinpoint pieces of code which may take too long as they are 'blocking' background tasks.
So that could be one of the reasons why things can't get processed in time.
But that still should not result in the wrong strings being processed.

I've seen it happen also a few times, but was never able to properly reproduce it.

clumsy-stefan commented:
ok, but I don't say that wrong strings are beeing processed, I just see that it has an influence if a web page is served as a whole or breaks in the mittle of the transfer (and has these strange RST's on a network layer).

I'll have a closer look at the timings then, but need to do a new version, as I don't include timing stats normally..

TD-er commented:
The problem is that as soon as memory allocation fails, all kinds of bad things can happen.
But I thought you mentioned you saw parts of the web page data being sent by the C009 HTTP POST calls?

clumsy-stefan commented:
Sorry, my not perfect english ;)

I just see that the HTTP (or json) page isn't interrupted more or less often depending on the settings I use in the controller... the issue is just that the web page only transmits partially and then stops.

clumsy-stefan commented:
<img width="1782" alt="image" src="https://user-images.githubusercontent.com/32835100/221172625-fbc4ec8f-7c13-41f6-8cfe-bb56c9939edd.png">
or a partially transferred page:
<img width="862" alt="image" src="https://user-images.githubusercontent.com/32835100/221173157-58f070fd-8323-464c-b75d-a1aeb4b5e9aa.png">

clumsy-stefan commented:
This node has a good reception and normally easily accessible, still a very long time to serve a (text-only) json page...

<img width="1775" alt="image" src="https://user-images.githubusercontent.com/32835100/221176404-33a67dbd-3b63-4446-93c7-fba132bd38c8.png">

TD-er commented:
That's indeed a very strange issue, as the JSON pages typically are being served quite fast.

Is it possible you're running out of "free slots" in the webserver?
I think it is set to 4 or 5.
And with poor wifi, it might be possible a connection is never closed.

clumsy-stefan commented:
> Is it possible you're running out of "free slots" in the webserver?

~where do I set this?~

Found it!

TD-er commented:
> Found it!

Where?
I was looking for it for you, but haven't found it yet

clumsy-stefan commented:
I guess these are compile time defines as described in [lwiopts.h](https://github.com/esp8266/Arduino/blob/master/tools/sdk/lwip2/include/lwipopts.h), this specific one I guess is `MEMP_NUM_TCP_PCB` (line 444).

EDIT: there are also a number of other things to tewak (buffers, memory, segments, etc.)... I'll try some of these ;)

clumsy-stefan commented:
One more observsation, sometimes I get the whole page, but I guess the connection is not closed. Sometimes I get part of the page, but still I do get the "end" of the page (see attached screenshot from a /json request)
<img width="647" alt="image" src="https://user-images.githubusercontent.com/32835100/221238272-27ae160a-54b4-41d1-8ae2-0ba3bc44ec45.png">
and sometimes only the CSS is missing and I get a unformated, but complete page...

TD-er commented:
The problem with the CSS is something I also have noticed quite a lot lately.

I think it may have to do with trying to serve the CSS  as a separate link. The idea is to let the ESP serve it once and let the browser cache it, but in reality the browser always still queries to see if it has changed and this may timeout or be missed.

Maybe we can add some 'defer' or something like that to the link to make sure the browser doesn't try to request both at the same time.

clumsy-stefan commented:
Hmm... That could explain why the node gets "overloaded" when it has some latency... when it then tries to send something to the controller (via HTTP) you're probably running out of network ressources (sockets, packets, fragments, whatever)...

Any possibility to temporarily "disable" this completely to verify if that would improve the situation?

TD-er commented:
> Any possibility to temporarily "disable" this completely to verify if that would improve the situation?

You could set the define to only serve static files from CDN.
That's the easiest way to change this as it doesn't require any other changes in the code.

clumsy-stefan commented:
these ones I guess:
```
#undef WEBSERVER_EMBED_CUSTOM_CSS
#undef WEBSERVER_CSS
#define WEBSERVER_INCLUDE_JS
#define WEBSERVER_FAVICON_CDN
#define WEBSERVER_USE_CDN_JS_CSS
```

However I don't think that changes anything as I just realize neither the controller send nor the /json get will use CSS.... That's only for the webpage I guess..

TD-er commented:
Yep, CSS and JS are only needed for serving the web pages.

clumsy-stefan commented:
After another fifty or so versions with different parameters, defines, changes, whatever, I'm close to give up searching. Some nodes just randomly stop sending data in the middle of the page, no matter if it's just the /json or /sysinfo or even the main page. Sometimes the page loads in no time and sometimes it just loads a couple hundreds of bytes..
The only indication I have is, that it seems to happen (more) on nodes with a bit a bad signal strength I can't really see any relation to anything else.

However, while digging I found one or two things which look a bit strange to me and probably could even also add to the issues I see... All of them in `src/DataStructs/Web_StreamingBuffer.cpp`

In the function `void Web_StreamingBuffer::endStream()`:
```
    if (buf.length() > 0) { sendContentBlocking(buf); }
    buf.clear();
    sendContentBlocking(buf);
```
After `buf.clear()` the buffer obviously is always empty, so why do a send again? This only sends an additional `\r\n` if I get this right... so either the conditional send or the second send can be omitted IMHO. At least I see no difference if I remove it.

Two lines further down for ESP8266 it says
```
    web_server.client().flush(100);
```
According to the documentation this would mean, that the `flush()` only waits 100ms for the client to receive all data. Especially on slow connections or high latencies this could probably cause issues.
In `Web_StreamingBuffer::sendHeaderBlocking()` the same `flush()` is called without a timeout, which seems to work.
Removing the timeout and just call `client().flush()` seems to improve the situation for me. I feel that I get less interrupts.

Generally it would probably a good idea to check for the returncode of `client().flush()` as per documentation it would return if the flush has been successful or not... if it was not, probably a second one would help?

And the a pure cosmetic one, again a few lines further down, the DEBUG level output is completely commented out. I guess this could be included with `#ifndef BUILD_NO_DEBUG` surrounding it.

But again, all of these are just heuristics and guesses, as I'm getting rather frustrated after days of searching... Probably i could also be a HW-Issue, which I don't want to beleive as it happens quite randomly on different nodes....

TD-er commented:
The timeout for the `flush()` could very well be related to what you're seeing.
The sending of an "empty" chunk with only /r/n is just what is needed to end the chunk transfers, since you can't give the expected size up front.

clumsy-stefan commented:
ok. thx.

and could it happen in `Web_StreamingBuffer::endStream()` that `buf.length()` is still bigger than `CHUNKED_BUFFER_SIZE` and therefore part of the buffer is not sent?

Also in `Web_StreamingBuffer::sendContentBlocking()`which is called from `Web_StreamingBuffer::endStream()` the buffer is cleared anyway, so why clear it again there?

TD-er commented:
> and could it happen in `Web_StreamingBuffer::endStream()` that `buf.length()` is still bigger than `CHUNKED_BUFFER_SIZE` and therefore part of the buffer is not sent?

Extremely unlikely, as all functions that fill this buffer check for the length.

TD-er commented:
> Also in `Web_StreamingBuffer::sendContentBlocking()`which is called from `Web_StreamingBuffer::endStream()` the buffer is cleared anyway, so why clear it again there?

Could be a left-over from previous code revisions.
But a clear() on an already cleared String is not doing anything.

clumsy-stefan committed:
Merge branch 'mega' into increase_chunked_timeout

clumsy-stefan committed:
[HTTP] Increase timeouts and smaller fixes

Remove the flush timeout of 100ms improving stability HTTP transfer for slow connections.

Minor code amendments for DEBUG Log and timeouts.

clumsy-stefan commented:
The commits I pushed are from the running version I currently use, which seems to improve reliability/stability for nodes with big latencies and/or weak signal strength.

Not sure though if this can/should be included, but it runs stable on ~40 nodes on my site (1M esp8285, 4M/16M esp8266, 16M ESP32)...

TD-er commented on src/src/DataStructs/Web_StreamingBuffer.cpp:
```diff
@@ -260,23 +260,21 @@ void Web_StreamingBuffer::endStream() {
         addLog(LOG_LEVEL_DEBUG, log);
         }
-      */
+#endif // ifndef BUILD_NO_DEBUG
```
The old state was that it was not even included in the 'normal' or 'max' builds, now you put it back in?

clumsy-stefan replied:
Yes, like with all other debug outputs so it can be enabled or disabled with a define at compile time, depending if you need the size or the debug output. I think it's done everywhere like this?

TD-er replied:
Point was, this was not needed in any build right now, as it is only for debugging purposes and if ever needed again a new build to enable it is perfectly fine :)

clumsy-stefan replied:
Which is true for all DEBUG output I guess ;) But I'm also perfectly fine without it 😁 I just stumbled over it and included it because I thought it's always done with the ifdef's instead of directly uncommenting it...

TD-er replied:
You don't get the point I think...
First it was commented out, so will never ever be built.
Now it will be built for those builds with `BUILD_NO_DEBUG` not being defined. (thus 'normal' and 'max' builds)

clumsy-stefan replied:
That's completely understood, but true for all `LOG_LEVEL_DEBUG` and `LOG_LEVEL_DEBUG_DEV` output statements, they are all included in the max und normal build because `BUILD_NO_DEBUG` is not defined. I just didn't see the point why this particular debug output should be handled differently... a few lines further down in `Web_StreamingBuffer::sendContentBlocking()` it's exactly the same, there is a `LOG_LEVEL_DEBUG_DEV` log statement which is built if not excluded...

But hey, dont' want to waste your time because of this, I'm happy to exclude it again completely... ;)

TD-er replied:
> But hey, dont' want to waste your time because of this, I'm happy to exclude it again completely... ;)

:)
Please do put it in comments again (the #ifndef check is fine)
Thing is, I've made commits to try to reduce the build size for less gain than this log entry.

clumsy-stefan committed:
Excluded debug output

As requested ;)

clumsy-stefan committed:
Merge remote-tracking branch 'origin/mega' into increase_chunked_timeout

clumsy-stefan commented:
One more thing I observed: when I change the `Extra WiFi scan loops` to something > 0, then the node goes offline after some time (~15-30min) and never reconnects anymore, even though it keeps running. Need to do a trace with a serial attached node though...

TD-er commented:
Yep, there is something funky going on with scanning while being connected.
So I guess you may need to check the checkbox to restart WiFi after a lost connection.

clumsy-stefan commented:
> So I guess you may need to check the checkbox to restart WiFi after a lost connection.

I don't really see a change. On nodes with bad reception they just become unavailable after some time. Reconnecting does not seem to work (on my nodes/environment).

TD-er commented:
Maybe, as a work-around, you can (re)set a timer in the rules based on the disconnect event and clear it on connect events.
If this timer expires, then force a reboot (don't use `reset`, just `reboot` ;) )

clumsy-stefan commented:
Yep, or use the `Connection Failure Threshold` value for reboots. Just wondering why this setting and code is in at all if not used or not working as expected...

TD-er commented:
The nr. of connection failed counter is probably only counting connect attempts to hosts, not to wifi.
I guess it would make sense to either have a 2nd one for counting WiFi connects, or let it count wifi connects too.

I think it should still be separated. For example for nodes on a moving vehicle, or those that only occasionally may connect to a hotspot.

clumsy-stefan commented:
Completely agree, just meant it at workaround for me.. Somewhere there already is a "connection failed" counter (at least on the sysinfo page"... 

BTW: would be cool if all of these additional values would alos be in the JSON page (eg. WD Count, last action before reboot, flash information, frequencies, etc.).... Just before you're getting bored and not know what do add 😁

clumsy-stefan committed:
Merge branch 'mega' into increase_chunked_timeout

clumsy-stefan committed:
Merge remote-tracking branch 'origin/mega' into increase_chunked_timeout

clumsy-stefan committed:
Merge branch 'mega' into increase_chunked_timeout

clumsy-stefan committed:
Merge remote-tracking branch 'origin/mega' into increase_chunked_timeout

clumsy-stefan commented:
I have this PR now running for over a week on 40 nodes of different types without issues. I feel the nodes are more reliable under weak signal or crowded wifi environment and I can't see any negative impact.

Therefore I'd vote for this PR to be included in the mega branch... IMHO...

TD-er commented:
> Therefore I'd vote for this PR to be included in the mega branch... IMHO...

OK :)
